### PR TITLE
feat: Twitter Post Action Implementation

### PR DESCRIPTION
**Pull request by 0xPBIT**
# Relates to:
N/A - Plugin implementation for Twitter posting functionality via action

# Risks
Low - This is a new plugin that adds Twitter posting capability. The main risks are:
- Failed tweets due to rate limiting or authentication issues
- Potential for tweets that are too long (mitigated by automatic truncation)
- No risk to existing functionality as this is an isolated plugin

# Background

## What does this PR do?
Adds a Twitter plugin that enables posting tweets through the agent-twitter-client package. Key features:
- Implements POST_TWEET action for sending tweets
- Handles Twitter authentication via environment variables
- Automatically truncates tweets that exceed character limits
- Provides detailed error handling and logging
- Supports dry run mode for testing

## What kind of change is this?
Features (non-breaking change which adds functionality)
- Adds new Twitter posting capability
- Uses the standalone agent-twitter-client package
- Implements proper error handling and character limit enforcement

# Documentation changes needed?
My changes require a change to the project documentation:
- Add Twitter plugin setup instructions
- Document required environment variables:
  - TWITTER_USERNAME
  - TWITTER_PASSWORD
  - TWITTER_EMAIL
  - TWITTER_DRY_RUN (optional)
  - TWITTER_2FA_SECRET (optional)

# Testing

## Where should a reviewer start?
1. Review `packages/plugin-twitter/src/actions/post.ts` for the main implementation
2. Check error handling in postTweet function
3. Review tweet composition logic in composeTweet function
4. Verify environment variable handling

## Detailed testing steps
1. Set up environment variables in `.env`:
```bash
TWITTER_USERNAME=your_username
TWITTER_PASSWORD=your_password
TWITTER_EMAIL=your_email
TWITTER_DRY_RUN=false
```

2. Test tweet posting:
```bash
# Start with dry run mode
TWITTER_DRY_RUN=true pnpm dev

# Test actual posting
TWITTER_DRY_RUN=false pnpm dev
```

3. Verify error handling:
- Try posting a tweet > 180 characters (should truncate)
- Try posting with invalid credentials (should show error)
- Try posting with missing credentials (should validate)

4. Check logging:
- Verify credential validation logs
- Check tweet content generation logs
- Verify error messages are descriptive

# Deploy Notes
- Requires agent-twitter-client package
- Environment variables must be set for Twitter authentication
- No database changes required

### CODE REVIEW

**Comment by 0xPBIT** [comment]
Several people have commented in the discord coders chat about twitter functionality from an action instead of the hard coded client loop. I've opened this PR just to show an example of how this can be done.

This is a small demo where I workshop a tweet with c3po agent to come up with a tweet. Then I can say "go tweet that" and the agent will put it on their twitter profile -- without using the client-twitter package.

Screenshot 2024-12-23 at 9 50 22 PM
Screenshot 2024-12-23 at 9 51 14 PM

File: packages/plugin-twitter/src/actions/post.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            count: 5,
+        });
+
+        const formattedHistory = formatMessages({
+            messages: recentMessages,
+            actors: state?.actorsData,
+        });
+
+        // Template for generating the tweet
+        const tweetTemplate = `
```
Comment (monilpat) [inline packages/plugin-twitter/src/actions/post.ts:32]: Please move this to its own file called templates.ts

File: packages/plugin-twitter/src/actions/post.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+5. Speaks from the perspective of ${runtime.character.name}
+
+Generate only the tweet text, no other commentary.`;
+
+        const context = await composeContext({
+            state,
+            template: tweetTemplate,
+        });
+
+        const tweetContent = await generateText({
```
Comment (monilpat) [inline packages/plugin-twitter/src/actions/post.ts:58]: Please use generate object which allows z objects and type safe approach to this which allows letting a character maxinum

File: packages/plugin-twitter/src/actions/post.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            runtime,
+            context,
+            modelClass: ModelClass.SMALL,
+            stop: ["\n"],
+        });
+
+        const trimmedContent = tweetContent.trim();
+
+        // Enforce character limit
+        if (trimmedContent.length > 180) {
```
Comment (monilpat) [inline packages/plugin-twitter/src/actions/post.ts:68]: Also this should be configurable for folks that have premium

File: packages/plugin-twitter/src/actions/post.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+async function postTweet(content: string): Promise<boolean> {
+    try {
+        const scraper = new Scraper();
+        const username = process.env.TWITTER_USERNAME;
+        const password = process.env.TWITTER_PASSWORD;
+        const email = process.env.TWITTER_EMAIL;
+        const twitter2faSecret = process.env.TWITTER_2FA_SECRET;
+
+        if (!username || !password) {
+            throw new Error(
```
Comment (monilpat) [inline packages/plugin-twitter/src/actions/post.ts:91]: elizaLogger

File: packages/plugin-twitter/src/actions/post.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        elizaLogger.log("Attempting to send tweet:", content);
+        const result = await scraper.sendTweet(content);
+
+        const body = await result.json();
+        elizaLogger.log("Tweet response:", body);
+
+        // Check for Twitter API errors
+        if (body.errors) {
+            const error = body.errors[0];
+            throw new Error(
```
Comment (monilpat) [inline packages/plugin-twitter/src/actions/post.ts:112]: please elizaLogger on all errors

File: packages/plugin-twitter/src/actions/post.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                return true;
+            }
+
+            return await postTweet(tweetContent);
+        } catch (error) {
+            elizaLogger.error("Error in post action:", error);
+            return false;
+        }
+    },
+    examples: [
```
Comment (monilpat) [inline packages/plugin-twitter/src/actions/post.ts:185]: would recommend more examples one per simile

File: packages/plugin-twitter/src/actions/post.ts (hunk 2 above)
Comment (monilpat) [reply packages/plugin-twitter/src/actions/post.ts:58]: Also please create a types file. Please review _examples/plugins for how best to write an plugin / action

**Review by monilpat — CHANGES_REQUESTED** [review]
Amazing work thanks for doing this please address comments

**Comment by 0xPBIT** [comment]
thanks for the review @monilpat this is my first time opening a PR so I'll address the changes.

File: packages/plugin-twitter/src/actions/post.ts (hunk 2 above)
Comment (0xPBIT) [reply packages/plugin-twitter/src/actions/post.ts]: I'll review this before opening any new PRs next time. Thanks for the suggestion.

**Pull request merged**